### PR TITLE
Spark Streaming on Yarn

### PR DESCRIPTION
**Pull request by mkuthan**

### CODE REVIEW

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 1]
```diff
@@ -0,0 +1,324 @@
...
+layout: post
+title: "Long running Spark Streaming jobs on YARN cluster"
+date: 2016-10-03
+comments: true
+categories: [spark, yarn, hdfs]
+---
+
+Long running Spark Streaming job, once submitted to the YARN cluster should run forever until it will be intentionally stopped.
+Any interruption introduces substantial processing delays and could drive to data loss or duplicates.
+Neither YARN nor Apache Spark have not been designed for running long running services.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: bez "not"

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 2]
```diff
@@ -0,0 +1,324 @@
+---
+layout: post
+title: "Long running Spark Streaming jobs on YARN cluster"
+date: 2016-10-03
+comments: true
+categories: [spark, yarn, hdfs]
+---
+
+Long running Spark Streaming job, once submitted to the YARN cluster should run forever until it will be intentionally stopped.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: long-running

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue \
+        --conf spark.speculation=true
+
+
+## Security
+
+On secured HDFS cluster, long runing Spark Streaming jobs fails due to Kerberos ticket expiration.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: long-running przez 2 n

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+    }
+
+In the Spark Streaming application, background thread should monitor ```/path/to/marker/file``` file, 
+and when the file disappears stop the context calling ```streamingContext.stop(stopSparkContext = true, stopGracefully = true)```.
+
+## Summary
+
+As you could see, configuration for mission critical Spark Streaming application deployed on YARN is quite complex. 
+It has been long, tedious and iterative learning process of all presented techniques by a few very smart devs. 
+But at the end, long running Spark Streaming applications deployed on highly utilized YARN cluster are extraordinary stable.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: extraordinarily

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown (hunk 2 above)
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: _A_ long-running Spark Streaming job ... to _a_ YARN cluster... until it is (present)

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 5]
```diff
@@ -0,0 +1,324 @@
+---
+layout: post
+title: "Long running Spark Streaming jobs on YARN cluster"
+date: 2016-10-03
+comments: true
+categories: [spark, yarn, hdfs]
+---
+
+Long running Spark Streaming job, once submitted to the YARN cluster should run forever until it will be intentionally stopped.
+Any interruption introduces substantial processing delays and could drive to data loss or duplicates.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: could lead?

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+Neither YARN nor Apache Spark have not been designed for running long running services.
+But they have been successfully adapted to growing needs of near real-time processing implemented as long running jobs.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: long-running

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+Successfully does not necessarily mean without technological challenges.
+
+This blog post summarizes my experiences in running mission critical, long running Spark Streaming jobs on secured YARN cluster.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: long-running, on _a_ secured

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+You will learn how to submit Spark Streaming application to the YARN cluster to avoid sleepless nights during on-call hours.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: to _a_ YARN...

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+
+## Fault tolerance
+
+In the YARN cluster mode Spark driver runs in the Application Master, the first container allocated by the application.
+This process is responsible for driving the application and requesting resources (Spark executors) from YARN.
+What is important, Application Master eliminates need for any another process that run during application lifecycle.
+Even if the edge Hadoop cluster node where the Spark Streaming job was submitted fails, the application stays unaffected.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: an edge node

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 10]
```diff
@@ -0,0 +1,324 @@
...
+In the YARN cluster mode Spark driver runs in the Application Master, the first container allocated by the application.
+This process is responsible for driving the application and requesting resources (Spark executors) from YARN.
+What is important, Application Master eliminates need for any another process that run during application lifecycle.
+Even if the edge Hadoop cluster node where the Spark Streaming job was submitted fails, the application stays unaffected. 
+
+To run Spark Streaming application in the cluster mode, ensure that the following parameters are given to spark-submit command:
+
+    spark-submit --master yarn --deploy-mode cluster
+
+Because Spark driver and Application Master share single JVM, any error in Spark driver stops our long running job.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: share a single JVM

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 11]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8
+ 
+## Performance
+
+When Spark Streaming application is submitted to the cluster, YARN queue where the job runs must be defined.
+I strongly recommend to use YARN Capacity Scheduler and separate queue for long running jobs.
+Without separate YARN queue your long running job will be preempted by massive Hive query sooner or later.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: without a, by a massive query

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 12]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability. 
+Processing time should stay below batch time to avoid delays. 
+I found that Spark speculative execution helps a lot, especially on busy cluster.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: I've found, on a busy

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 13]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability. 
+Processing time should stay below batch time to avoid delays. 
+I found that Spark speculative execution helps a lot, especially on busy cluster. 
+Batch processing times are much more stable than when speculative execution is disabled.
+Unfortunately speculative mode can be enabled only if Spark actions are idempotent (like saving results on HDFS but not to Elastic).
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: darowałbym sobie przykłady, bo w obu przypadkach można to zrobić idempotentnie lub nie

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown (hunk 3 above)
Comment (arciuchm) [reply source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: on a

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 14]
```diff
@@ -0,0 +1,324 @@
...
+         --conf spark.hadoop.fs.hdfs.impl.disable.cache=true \
+         --conf spark.driver.extraJavaOptions=-Dlog4j.configuration=file:log4j.properties \
+         --conf spark.executor.extraJavaOptions=-Dlog4j.configuration=file:log4j.properties \
+         --files /path/to/log4j.properties:/path/to/metrics.properties
+
+Spark publishes tons of metrics from driver and executors.
+If I have to choose the most important one, it would be the last received batch records.
+When ```StreamingMetrics.streaming.lastReceivedBatch_records == 0``` it probably means that Spark Streaming job has been stopped or failed.
+
+Another important metrics are listed below:
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: Other (Another to coś pojedynczego)

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 15]
```diff
@@ -0,0 +1,324 @@
...
+
+When total delay is greater then batch interval, latency of the processing pipeline increases.
+
+    driver.StreamingMetrics.streaming.lastCompletedBatch_totalDelay
+
+When number of active tasks is lower then ```number of executors * number of cores```, allocated resources are not fully utilized.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: than

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 16]
```diff
@@ -0,0 +1,324 @@
...
+         --conf spark.executor.extraJavaOptions=-Dlog4j.configuration=file:log4j.properties \
+         --files /path/to/log4j.properties:/path/to/metrics.properties
+
+Spark publishes tons of metrics from driver and executors.
+If I have to choose the most important one, it would be the last received batch records.
+When ```StreamingMetrics.streaming.lastReceivedBatch_records == 0``` it probably means that Spark Streaming job has been stopped or failed.
+
+Another important metrics are listed below:
+
+When total delay is greater then batch interval, latency of the processing pipeline increases.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: than

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 17]
```diff
@@ -0,0 +1,324 @@
...
+    [0-9]*.jvm.pools.G1-Survivor-Space.used
+    [0-9]*.jvm.pools.G1-Eden-Space.used
+
+How much time is spent on GC on Spark executors.
+
+    [0-9]*.jvm.G1-Old-Generation.time
+    [0-9]*.jvm.G1-Young-Generation.time
+
+
+When You configure first Grafana dashboard for Spark job, perhaps the first question will emerge:
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: you

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 18]
```diff
@@ -0,0 +1,324 @@
...
+
+The first attempt to solve graceful shutdown issue was to call Spark streaming context stop method in shutdown hook.
+
+``` scala
+sys.addShutdownHook {
+    streamingContext.stop(stopSparkContext = true, stopGracefully = true)
+}
+```
+
+Disappointingly shutdown hook is called too late to finish started batch and Spark application is killed almost immediately.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: a hook tu i linijkę niżej

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 19]
```diff
@@ -0,0 +1,324 @@
...
+Moreover there is no guarantee that shutdown hook will be called by JVM at all.
+
+When I'm writing this blog post the only confirmed way to shutdown gracefully Spark Streaming application on YARN
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: "At the time of writing"

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 20]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8
+ 
+## Performance
+
+When Spark Streaming application is submitted to the cluster, YARN queue where the job runs must be defined.
+I strongly recommend to use YARN Capacity Scheduler and separate queue for long-running jobs.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: recommend using zamiast to use

Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: submitting long-running jobs to separate queue ?

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 21]
```diff
@@ -0,0 +1,324 @@
...
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: Another important issue for Spark Streaming job is keeping processing time(micro batch duration) stable and highly predictable

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 22]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability. 
+Processing time should stay below batch time to avoid delays. 
+I've found that Spark speculative execution helps a lot, especially on busy cluster. 
+Batch processing times are much more stable than when speculative execution is disabled.
+Unfortunately speculative mode can be enabled only if Spark actions are idempotent.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown:83]: Although, not idempotent actions such as writing file to HDFS works well. Suppose that we have 2 running tasks, which are both attempting to write the same file (under identical path) to HDFS, the first one that succeeds will save the file, and second will be skipped or failed without any damage done to job.

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 23]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important performance factor for Spark Streaming job is processing time predictability. 
+Processing time should stay below batch time to avoid delays. 
+I've found that Spark speculative execution helps a lot, especially on busy cluster.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: The idea behind speculative execution is quite simple. For example, let us suppose that we have operation which runs on multiple executors and one of them has a lag. After some time Spark will notice that this executor is running slower than others and will try to do same operation on different and better performing executor.

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 24]
```diff
@@ -0,0 +1,324 @@
...
+         --conf spark.speculation=true \
+         --principal user/hostname@domain \
+         --keytab /path/to/foo.keytab \
+         --conf spark.hadoop.fs.hdfs.impl.disable.cache=true \
+         --conf spark.driver.extraJavaOptions=-Dlog4j.configuration=file:log4j.properties \
+         --conf spark.executor.extraJavaOptions=-Dlog4j.configuration=file:log4j.properties \
+         --files /path/to/log4j.properties:/path/to/metrics.properties
+
+Spark publishes tons of metrics from driver and executors.
+If I have to choose the most important one, it would be the last received batch records.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: If I were

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 25]
```diff
@@ -0,0 +1,324 @@
...
+
+When number of active tasks is lower than ```number of executors * number of cores```, allocated resources are not fully utilized.
+
+    executor.threadpool.activeTasks
+
+How much RAM is used for RDD cache.
+
+    driver.BlockManager.memory.memUsed_MB
+
+When there is not enough RAM for RDD cache, how much data has been spilled to disk.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: how much data should be

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 26]
```diff
@@ -0,0 +1,324 @@
...
+    [0-9]*.jvm.pools.G1-Survivor-Space.used
+    [0-9]*.jvm.pools.G1-Eden-Space.used
+
+How much time is spent on GC on Spark executors.
+
+    [0-9]*.jvm.G1-Old-Generation.time
+    [0-9]*.jvm.G1-Young-Generation.time
+
+
+When you configure first Grafana dashboard for Spark job, perhaps the first question will emerge:
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: When you are done configurating. Albo After you have configured.

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 27]
```diff
@@ -0,0 +1,324 @@
...
+{% img /images/blog/spark_job_monitoring.png [spark job monitoring] %}
+
+If Spark application is restarted frequently, metrics for old, already finished runs should be deleted from Graphite.
+Because Graphite does not compact inactive metrics, old metrics slow down Graphite itself and Grafana queries.
+
+## Graceful stop
+
+The last puzzle element is how to stop Spark Streaming application deployed on YARN in a graceful way.
+The standard method for stopping (or rather killing) YARN application is using a command ```yarn application -kill [applicationId]```.
+And this command stops the Spark Streaming application but the application might be killed in the middle of the batch.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: przecinek przed but chyba

Comment (pguzik) [reply source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: but the application might already be killed

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 28]
```diff
@@ -0,0 +1,324 @@
...
+So if the job reads data from Kafka, save processing results on HDFS and finally commit Kafka offsets
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: saves a potem commits

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 29]
```diff
@@ -0,0 +1,324 @@
...
+    streamingContext.stop(stopSparkContext = true, stopGracefully = true)
+}
+```
+
+Disappointingly a shutdown hook is called too late to finish started batch and Spark application is killed almost immediately.
+Moreover there is no guarantee that a shutdown hook will be called by JVM at all.
+
+At the time of writing this blog post the only confirmed way to shutdown gracefully Spark Streaming application on YARN
+is to notify somehow the application about planned shutdown, and then stop streaming context programmatically (but not from shutdown hook).
+Command ```yarn application -kill``` should be used only as a last resort if notified application did not stop after defined timeout.
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown:304]: is notifying the application

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 30]
```diff
@@ -0,0 +1,324 @@
...
+
+The application can be notified about planned shutdown using marker file on HDFS (the easiest way),
```
Comment (pguzik) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown:306]: Marker file could be stored under path like /var/run/spark_application_name. Marker file works similiar to PID file in UNIX based systems.

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 31]
```diff
@@ -0,0 +1,324 @@
...
+    spark-submit --master yarn --deploy-mode cluster \
+        --conf spark.yarn.maxAppAttempts=4 \
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8
+ 
+## Performance
+
+When Spark Streaming application is submitted to the cluster, YARN queue where the job runs must be defined.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: When a Spark Streaming application

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 32]
```diff
@@ -0,0 +1,324 @@
...
+        --conf spark.yarn.am.attemptFailuresValidityInterval=1h \
+        --conf spark.yarn.max.executor.failures={8 * num_executors} \
+        --conf spark.yarn.executor.failuresValidityInterval=1h \
+        --conf spark.task.maxFailures=8 \
+        --queue realtime_queue
+
+
+Another important issue for Spark Streaming job is keeping processing time stable and highly predictable.
+Processing time should stay below batch duration to avoid delays. 
+I've found that Spark speculative execution helps a lot, especially on busy cluster.
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: on a busy cluster

File: source/_posts/2016-09-30-spark-streaming-on-yarn.markdown [hunk 33]
```diff
@@ -0,0 +1,324 @@
...
+         --queue realtime_queue \
+         --conf spark.speculation=true \
+         --principal user/hostname@domain \
+         --keytab /path/to/foo.keytab \
+         --conf spark.hadoop.fs.hdfs.impl.disable.cache=true
+
+
+## Logging
+
+The easiest way for Spark jobs to access application logs is to configure Log4j console appender,
```
Comment (arciuchm) [inline source/_posts/2016-09-30-spark-streaming-on-yarn.markdown]: "The easiest way to access Spark job logs..." - teraz brzmi jak by to Spark szukał swoich logów

**Pull request merged**